File: LEETCODE/LEETCODE-JAVA/205.java
// 205. Isomorphic Strings
// https://leetcode.com/problems/isomorphic-strings/

class Solution {
    public boolean isIsomorphic(String s, String t) {
        int m=s.length(), n=t.length();

        if(m!=n){
            return false;
        }

        HashMap<Character,Character> sMap= new HashMap<>();
        HashMap<Character,Character> tMap= new HashMap<>();

        for(int i=0;i<m;i++){
            char x=s.charAt(i), y=t.charAt(i);

            if(sMap.containsKey(x) && sMap.get(x)!=y){
                return false;
            }
            if(tMap.containsKey(y) && tMap.get(y)!=x){
                return false;
            }

            sMap.put(x,y);
            tMap.put(y,x);

        }

        return true;
    }
}

File: LEETCODE/LEETCODE-PYTHON/205.py
# 205. Isomorphic Strings
# https://leetcode.com/problems/isomorphic-strings/
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        m=len(s)
        n=len(t)

        if m!=n:
            return False

        s_map={}
        t_map={}

        for x, y in zip(s,t):
            if x in s_map and s_map[x]!=y:
                return False

            if y in t_map and t_map[y]!=x:
                return False
        
            s_map[x]=y
            t_map[y]=x

        return True